test/fw/ptl/utils/pbs_dshutils.py

```diff
@@ -2069,55 +2071,66 @@ def create_temp_dir(self, hostname=None, suffix='', prefix='PtlPbs',
         :type suffix: str
         :param prefix: the directory name will begin with this prefix
         :type prefix: str
-        :param dir: the directory will be created in this directory
-        :type dir: str or None
-        :param uid: Optional username or uid of temp directory owner
-        :param gid: Optional group name or gid of temp directory
+        :param dirname: the directory will be created in this directory
+        :type dirname: str or None
+        :param asuser: Optional username of temp directory owner
+        :type asuser: str
+        :param asgroup: Optional group name of temp directory
                     group owner
+        :type asgroup: str
         :param mode: Optional mode bits to assign to the temporary
                      directory
+        :type mode: octal integer
         :param level: logging level, defaults to INFOCLI2
         """
+        current_user_info = self.get_id_info(self.get_current_user())
+        uid = current_user_info['uid']
+        if asuser is not None:
+            uid = PbsUser.get_user(asuser).uid
+        if asgroup is not None:
+            gid = PbsGroup.get_group(asgroup).gid
+        else:
+            gid = None
         # create a temp dir as current user
         tmpdir = tempfile.mkdtemp(suffix, prefix)
+        # By default mkdtemp creates dir according to umask.
+        # To create dir as different user first change the dir
+        # permission to 0755 so that other user has read permission
+        self.chmod(path=tmpdir, mode=0o755)
         if dirname is not None:
             dirname = str(dirname)
             self.run_copy(hostname, src=tmpdir, dest=dirname, runas=asuser,
-                          recursive=True,
-                          preserve_permission=False, level=level, sudo=sudo)
+                          recursive=True, gid=gid, uid=uid,
+                          level=level)
+            self.chmod(hostname, path=dirname, mode=mode, runas=asuser)
+
             tmpdir = dirname + tmpdir[4:]
 
         # if temp dir to be created on remote host
         if not self.is_localhost(hostname):
             if asuser is not None:
-                # by default mkstemp creates dir with 0600 permission
-                # to create dir as different user first change the dir
-                # permission to 0644 so that other user has read permission
-                self.chmod(path=tmpdir, mode=0o755)
                 # copy temp dir created on local host to remote host
                 # as different user
-                self.run_copy(hostname, src=tmpdir, dest=tmpdir, runas=asuser,
-                              recursive=True,
-                              preserve_permission=False, level=level)
+                self.run_copy(hostname, src=tmpdir, dest=tmpdir,
+                              recursive=True, uid=uid, gid=gid,
+                              level=level)
+                self.chmod(hostname, path=tmpdir, mode=mode, runas=asuser)
```

Comment:
regardless of whether asuser was provided or not, don't we want to do chmod if 'mode' was provided? I know this is not your code, but I'm just wondering if the earlier logic for setting was mode was also incorrect, shouldn't it be outside the if asuser block?